File: app/src/main/java/dev/favourdevlabs/cleanthes/utils/DateUtils.java
package dev.favourdevlabs.cleanthes.utils;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class DateUtils {

    private static final String FORMAT_FULL = "MMM d, yyyy h:mm a";
    private static final String FORMAT_SHORT = "MMM d, yyyy";

    private DateUtils() {
    }

    public static String formatFull(long epochMillis) {
        SimpleDateFormat sdf = new SimpleDateFormat(FORMAT_FULL, Locale.getDefault());
        return sdf.format(new Date(epochMillis));
    }

    public static String formatShort(long epochMillis) {
        SimpleDateFormat sdf = new SimpleDateFormat(FORMAT_SHORT, Locale.getDefault());
        return sdf.format(new Date(epochMillis));
    }

    public static String formatRelative(long epochMillis) {
        long now = System.currentTimeMillis();
        long delta = now - epochMillis;

        long seconds = delta / 1000;
        long minutes = seconds / 60;
        long hours = minutes / 60;
        long days = hours / 24;

        if (seconds < 60)
            return "just now";
        if (minutes < 60)
            return minutes + (minutes == 1 ? " minute ago" : " minutes ago");
        if (hours < 24)
            return hours + (hours == 1 ? " hour ago" : " hours ago");
        if (days < 7)
            return days + (days == 1 ? " day ago" : " days ago");

        return formatShort(epochMillis);
    }
}

File: core/common/src/main/java/dev/favourdevlabs/cleanthes/common/DateUtils.kt
package dev.favourdevlabs.cleanthes.common

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

private const val FORMAT_FULL  = "MMM d, yyyy h:mm a"
private const val FORMAT_SHORT = "MMM d, yyyy"

fun formatFull(epochMillis: Long): String =
    SimpleDateFormat(FORMAT_FULL, Locale.getDefault()).format(Date(epochMillis))

fun formatShort(epochMillis: Long): String =
    SimpleDateFormat(FORMAT_SHORT, Locale.getDefault()).format(Date(epochMillis))

fun formatRelative(epochMillis: Long): String {
    val delta   = System.currentTimeMillis() - epochMillis
    val seconds = delta / 1000
    val minutes = seconds / 60
    val hours   = minutes / 60
    val days    = hours   / 24

    return when {
        seconds < 60  -> "just now"
        minutes < 60  -> "$minutes ${if (minutes == 1L) "minute ago" else "minutes ago"}"
        hours   < 24  -> "$hours ${if (hours   == 1L) "hour ago"   else "hours ago"}"
        days    < 7   -> "$days ${if (days     == 1L) "day ago"    else "days ago"}"
        else          -> formatShort(epochMillis)
    }
}
